### ISSUE TITLE
Downloads problem, where to go next? 

### ISSUE THREAD

**Issue by chris-sloan** [google/ExoPlayer#10225]
We have a VOD streaming app which allows users to donwload MPEG-DASH WIDEVINE content. 

If the user launches the app and immediately begins a download then the download will happen correctly and all the data will be downloaded and playback on the downloaded data can be achieved. 
But if the user initiates online playback in some way prior to initiating a download (could be entirely separate content) then the download starts as expected but almost immediately is being reported as completed. 

I've tried to trace the code back to see where and why this happens, but can only get so far, and am looking for some pointers as to where I should next direct my attention. 

In `DownloadManager.java`, I can see that all download updates are sent with a key of `MSG_DOWNLOAD_UPDATE` and there are 4 different locations where this is called. Eventually I can trace back to `DownloadManager.InternalHandler.handleMessage` method but then I can't find what is sending the message in the first instance. 
<img width="1227" alt="Screenshot 2022-04-28 at 13 07 59" src="https://user-images.githubusercontent.com/9352594/165749037-dddf9b39-b08a-489f-aef1-24b9d9bcea74.png">

Where could I go from here to further diagnose why the download is flagged as being completed? 

Are there any other things that spring to mind with a problem of this sort that perhaps I haven't thought of? 

I searched past issues for anything that resembled this problem but to no avail. 

Thanks

**Comment by chris-sloan** [comment]
Bumping this in the hope that someone can point me in the right direction.

Where can I find out what causes the download manager to report it has complete prematurely?

**Comment by tonihei** [comment]
Could you provide us with detailed reproduction steps for this problem? Based on your description it sounds as if something might be wrong, but to actually debug this we'd need very concrete steps to replicate the problem either by changes that need to be done to the demo app, or by providing a small sample project. 

> someone can point me in the right direction.

If you want to debug this yourself, you should be able to find the sender of this message by looking for the `MSG_TASK_STOPPED` constant in the code (L1373 most likely). 

> But if the user initiates online playback in some way prior to initiating a download (could be entirely separate content) then the download starts as expected but almost immediately is being reported as completed.

Depending on your setup, it may be expected that the download is stopped/paused while playback is ongoing to not compete for the same network resource. It's worth checking again what the state of the download is exactly (is really "completed", or is it just the current task that stopped?)

**Comment by chris-sloan** [comment]
Thanks for the response.

I worked out that there were some keys being retained from the previous playback session, and so the download was only actually downloading the subtitles text content as a result. Because it happened almost immediately, it looked like the download was failing prematurely but in fact was completing correctly. 

By ensuring those keys weren't retained following the completion of the playback session, I could successfully download the content once again. 

Closing now.
